crates/lint/src/linter.rs

```diff
@@ -0,0 +1,275 @@
+use clap::ValueEnum;
+use core::fmt;
+use foundry_compilers::Language;
+use solar_ast::ast::Span;
+use std::{
+    collections::{BTreeMap, HashMap},
+    error::Error,
+    hash::Hash,
+    ops::{Deref, DerefMut},
+    path::PathBuf,
+};
+use yansi::Paint;
+
+/// Trait representing a generic linter for analyzing and reporting issues in smart contract source
+/// code files. A linter can be implemented for any smart contract language supported by Foundry.
+///
+/// # Type Parameters
+///
+/// - `Language`: Represents the target programming language. Must implement the [`Language`] trait.
+/// - `Lint`: Represents the types of lints performed by the linter. Must implement the [`Lint`]
+///   trait.
+/// - `LinterError`: Represents errors that can occur during the linting process.
+///
+/// # Required Methods
+///
+/// - `lint`: Scans the provided source files and returns a [`LinterOutput`] containing categorized
+///   findings or an error if linting fails.
+pub trait Linter: Send + Sync + Clone {
+    type Language: Language;
+    type Lint: Lint + Ord;
+    type LinterError: Error + Send + Sync + 'static;
+
+    fn lint(&self, input: &[PathBuf]) -> Result<LinterOutput<Self>, Self::LinterError>;
+}
+
+pub struct ProjectLinter<L>
+where
+    L: Linter,
+{
+    pub linter: L,
+}
+
+impl<L> ProjectLinter<L>
+where
+    L: Linter,
+{
+    pub fn new(linter: L) -> Self {
+        Self { linter }
+    }
+
+    pub fn lint(self, input: &[PathBuf]) -> eyre::Result<LinterOutput<L>> {
+        Ok(self.linter.lint(input)?)
+    }
+}
+
+#[derive(Default)]
+pub struct LinterOutput<L: Linter>(pub BTreeMap<L::Lint, Vec<SourceLocation>>);
+
+impl<L: Linter> LinterOutput<L> {
+    pub fn new() -> Self {
+        Self(BTreeMap::new())
+    }
+}
+
+impl<L: Linter> Deref for LinterOutput<L> {
+    type Target = BTreeMap<L::Lint, Vec<SourceLocation>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<L: Linter> DerefMut for LinterOutput<L> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<L: Linter> Extend<(L::Lint, Vec<SourceLocation>)> for LinterOutput<L> {
+    fn extend<T: IntoIterator<Item = (L::Lint, Vec<SourceLocation>)>>(&mut self, iter: T) {
+        for (lint, findings) in iter {
+            self.0.entry(lint).or_default().extend(findings);
+        }
+    }
+}
+
+impl<L: Linter> fmt::Display for LinterOutput<L> {
```

Comment:
What's the reason of collecting results in a report and printing it manually rather than using the solar diagnostics system? It's way more flexible and it's the same structure as diagnostics in Rust

Reply:
Sounds good, I was not aware of the solar diagnostics system, I can update to use this.